File: 79. Word Search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not board or not word:
            return False
        
        visited = set()
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    if self.dfs(board, visited, i, j, word, 0):
                        return True
        
        return False
        
    def dfs(self, board, visited, i, j, word, k):
        if k == len(word):
            return True
        
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or (i, j) in visited or word[k] != board[i][j]:
            return False
        
        visited.add((i, j))
        res = self.dfs(board, visited, i-1, j, word, k+1) or \
              self.dfs(board, visited, i, j-1, word, k+1) or \
              self.dfs(board, visited, i+1, j, word, k+1) or \
              self.dfs(board, visited, i, j+1, word, k+1) 
        
        visited.remove((i, j))
        return res
    
        

File: 79. Word Search/solution.java
class Solution {
    public boolean exist(char[][] board, String word) {        
        //edge cases
        if (board.length == 0 || board[0].length == 0) {
            return false;
        }
        
        if (word.length() == 0) {
            return false;
        }
        
        // allocate visited grid
        int[][] visited = new int[board.length][board[0].length];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                visited[i][j] = 0;
            }
        }
        
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (word.charAt(0) == board[i][j]) {
                    // System.out.printf("%d, %d, %c \n", i, j, board[i][j]);
                    if (dfs(board, i, j, word, 0, visited)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    
    public boolean dfs(char[][] board, int i, int j, String word, int k, int[][] visited) {
        if (i < 0 || j < 0 || i >= board.length || j >= board[0].length || 
            board[i][j] != word.charAt(k) || 
            visited[i][j] == 1) {
            return false;
        }
        
        if (k == word.length()-1) {
            return true;
        }
        
        visited[i][j] = 1;
        boolean found = dfs(board, i-1, j, word, k+1, visited) ||
            dfs(board, i, j-1, word, k+1, visited) ||
            dfs(board, i+1, j, word, k+1, visited) ||
            dfs(board, i, j+1, word, k+1, visited);
        visited[i][j] = 0;
        return found;
    }
}
